### PR TITLE
Use assim_freq to configure "offset" in dependencies

### LINKED ISSUE #2038: use assim_freq in config.base to configure "offset" in dependencies
### What new functionality do you need?

use `assim_freq` in config.base to configure "offset" in dependencies.
This is the very first step to develop a workflow cycled more frequently than 6-hourly.

### What are the requirements for the new functionality?

Change the hard-coded "06:00:00" to a variable whose content is determined by the "`assim_freq`" setting in config.base

### Acceptance Criteria

The generated rotoco xml files are the same between new and old versions for a 6-hourly workflow (excluding the different timestamps about when the xml files are generated)

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by guoqing-noaa**
# Description
Use the assim_freq from config.base to configure "offset" in dependencies instead of hard-coding them.

Resolves #2038

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?

Tested on Hera:
Compared two rocoto xml files generated separately by the latest develop branch and this PR and they are the same except the difference in the timestamp information about when the xml files were generated.

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: workflow/rocoto/gfs_tasks.py
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: I would suggest adding the following in `class Tasks` after L35 in [tasks.py](https://github.com/NOAA-EMC/global-workflow/blob/75269e4bb9764b81589efc7d703825b80e74c8f5/workflow/rocoto/tasks.py#L35)
```python3
self._base['cycling_interval'] = to_timedelta(self._base["assim_freq"])
```
(The above might need to be made a string of `HH:00:00`)
and then use `[-]self._base.cycling_interval` where necessary instead of calculating `myoffset` each time.

Comment (guoqing-noaa) [reply workflow/rocoto/gfs_tasks.py]: Thank @aerorahul for the great suggestion! I will make this change soon.

Comment (guoqing-noaa) [reply workflow/rocoto/gfs_tasks.py]: @aerorahul I tried wxflow.timetools.to_timedelta(), it will need something as follows:
```
from wxflow import to_timedelta
self._base['cycling_interval'] =str(to_timedelta(str(self._base['assim_freq'])+'H'))

```
This looks a little bit more complicated than using the str.format().
Hope I did not miss anything.

I will add a dictionary entry as suggested to avoid calculating `myoffset `many times but will use str.format(). Thanks!

File: workflow/rocoto/gfs_cycled_xml.py
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/gfs_cycled_xml.py]:
```suggestion
        interval = f'{self._base["assim_freq"]:02n}:00:00'
```

File: workflow/rocoto/gfs_tasks.py
Comment (WalterKolczynski-NOAA) [reply workflow/rocoto/gfs_tasks.py]: Current python prefers using f-strings over `.format()` or concatenation when possible:
```
from wxflow import to_timedelta
self._base['cycling_interval'] =str(to_timedelta(f'{self._base["assim_freq"]}H'))
```

wxflow also has a function to convert a timedelta to HH:MM:SS:
```
from wxflow import to_timedelta, timedelta_to_HMS
self._base['cycling_interval'] = timedelta_to_HMS(to_timedelta(f'{self._base["assim_freq"]}H'))
```

However, I don't think you need to even bother converting to a string (the default casting of a timedelta to a string is the proper format), so this probably works:
```
from wxflow import to_timedelta
self._base['cycling_interval'] = to_timedelta(f'{self._base["assim_freq"]}H')
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: workflow/rocoto/gfs_tasks.py
Comment (aerorahul) [reply workflow/rocoto/gfs_tasks.py]: My apologies @guoqing-noaa 
I should have been more specific in my change request.
@WalterKolczynski-NOAA thank you for clarifying and providing the exact methods that perform the operations.

File: workflow/rocoto/tasks.py
Comment (aerorahul) [inline workflow/rocoto/tasks.py]: One should not define a specific use variable so high in the class.
```suggestion
        self._base['cycle_interval'] =  to_timedelta(f"{self._base['assim_freq']}H") 
```

File: workflow/rocoto/gfs_tasks.py
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: create the `minus` here with appropriate methods.
```suggestion
        dep_dict = {'type': 'metatask', 'name': 'gdaspost', 'offset': f"-{timedelta_to_HMS(self._base['cycle_interval'])}"}
```

File: workflow/rocoto/gfs_cycled_xml.py
Comment (aerorahul) [inline workflow/rocoto/gfs_cycled_xml.py]: `interval` should be, but currently is not, a `timedelta` object, just like `sdate` and `edate` are `datetime` objects.
If you can make that change, here and promote it elsewhere in this method, it would be the proper use of the datatype.  But, I understand if you would rather leave that to the developers.

**Review by aerorahul — CHANGES_REQUESTED** [review]
A few changes are requested in `tasks.py` and `gfs_tasks.py`.  
Not sure if corresponding to `gfs_tasks.py` changes are also needed in `gefs_tasks.py`.  @WalterKolczynski-NOAA?
Also a note on `gfs_cycled_xml.py`

**Comment by guoqing-noaa** [comment]
Thanks @WalterKolczynski-NOAA and @aerorahul for the great help and detailed suggestions. 
I committed new changes based on the review.

File: workflow/rocoto/gfs_cycled_xml.py
Comment (guoqing-noaa) [reply workflow/rocoto/gfs_cycled_xml.py]: @aerorahul Great suggestion! I would hope I can do so right now. However, I am not very familiar with this part. I hope I will learn more and more about this and be able to do more in the future. Thanks for the understanding!

**Review by aerorahul — APPROVED** [review]
looks good to me.
thank you for incorporating the suggestions.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests.

File: workflow/applications/applications.py
Comment (WalterKolczynski-NOAA) [inline workflow/applications/applications.py:176]: @aerorahul Any reason these keys are strings instead of just using ints and avoiding the cast in L179?

Comment (aerorahul) [reply workflow/applications/applications.py:176]: They can be `int`.

Comment (WalterKolczynski-NOAA) [reply workflow/applications/applications.py:176]: Also, looks like the `0` key (cycling only) was removed and needs to be added back in unless some if upstream stops this from being called.

Comment (aerorahul) [reply workflow/applications/applications.py:176]: I think I don't call this method at all if `gfs_cyc == 0`.  If I missed someplace, please let me know.

Comment (aerorahul) [reply workflow/applications/applications.py:176]: see the updated `gfs_cycled.py` line 253

Comment (guoqing-noaa) [reply workflow/applications/applications.py:176]: I can update 1, 2, 4 from strings to ints if that's desired. Thanks!

**Comment by WalterKolczynski-NOAA** [comment]
@guoqing-noaa Can you pull in develop and handle any conflicts? Then we'll rerun the tests and merge.

**Comment by guoqing-noaa** [comment]
> @guoqing-noaa Can you pull in develop and handle any conflicts? Then we'll rerun the tests and merge.

@WalterKolczynski-NOAA I merged the latest develop branch. I think it is ready for rerunning the test. Thanks!

**Comment by guoqing-noaa** [comment]
@WalterKolczynski-NOAA Sorry that the bug fix was lost when merging the latest develop branch. Now I added it back. Could you restart the tests?  Thanks!

**Comment by guoqing-noaa** [comment]
Thank you, @WalterKolczynski-NOAA and @aerorahul for your great help on this PR!

**Pull request merged**
